builder/docker/docker.go

```diff
@@ -0,0 +1,337 @@
+				log.Errf(c, "Failed to build docker image: %s", err)
+				m.Lock()
+				fails++
+				m.Unlock()
+			}
+			wg.Done()
+		})
+
+	}
+
+	// Number of failures.
+	c.Put("ParallelBuild.failN", fails)
+
+	return &wg, nil
+}
+
+// Waiter describes a thing that can wait.
+//
+// It does not bring you food. I should know. I tried.
+type Waiter interface {
+	Wait()
+}
+
+// Wait waits for a sync.WaitGroup to finish.
+//
+// Params:
+// 	- wg (Waiter): The thing to wait for.
+// 	- msg (string): The message to print when done. If this is empty, nothing is sent.
+// 	- waiting (string): String to tell what we're waiting for. If empty, nothing is displayed.
+// 	- failures (int): The number of failures that occurred while waiting.
+//
+// Returns:
+//  Nothing.
+func Wait(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	ok, missing := p.RequiresValue("wg")
+	if !ok {
+		return nil, &cookoo.FatalError{"Missing required fields: " + strings.Join(missing, ", ")}
+	}
+	wg := p.Get("wg", nil).(Waiter)
+	msg := p.Get("msg", "").(string)
+	fails := p.Get("failures", 0).(int)
+	waitmsg := p.Get("waiting", "").(string)
+
+	if len(waitmsg) > 0 {
+		log.Info(c, waitmsg)
+	}
+
+	wg.Wait()
+	if len(msg) > 0 {
+		log.Info(c, msg)
+	}
+
+	if fails > 0 {
+		return nil, fmt.Errorf("There were %d failures while waiting.", fails)
+	}
+	return nil, nil
+}
+
+// BuildImage builds a docker image.
+//
+// Essentially, this executes:
+// 	docker build -t TAG PATH
+//
+// Params:
+// 	- path (string): The path to the image. REQUIRED
+// 	- tag (string): The tag to build.
+func BuildImage(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	path := p.Get("path", "").(string)
+	tag := p.Get("tag", "").(string)
+
+	log.Infof(c, "Building docker image %s (tag: %s)", path, tag)
+
+	return buildImg(c, path, tag)
+}
+
+func buildImg(c cookoo.Context, path, tag string) ([]byte, error) {
+	dargs := []string{"build"}
+	if len(tag) > 0 {
+		dargs = append(dargs, "-t", tag)
+	}
+
+	dargs = append(dargs, path)
+
+	out, err := exec.Command("docker", dargs...).CombinedOutput()
+	if len(out) > 0 {
+		log.Infof(c, "Docker: %s", out)
+	}
+	return out, err
+}
+
+// Push pushes an image to the registry.
+//
+// This finds the appropriate registry by looking it up in etcd.
+//
+// Params:
+// - client (etcd.Getter): Client to do etcd lookups.
+// - tag (string): Tag to push.
+//
+// Returns:
+//
+func Push(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	// docker tag deis/slugrunner:lastest HOST:PORT/deis/slugrunner:latest
+	// docker push HOST:PORT/deis/slugrunner:latest
+	client := p.Get("client", nil).(etcd.Getter)
+
+	host, err := client.Get("/deis/registry/host", false, false)
+	if err != nil {
+		return nil, err
+	}
+	port, err := client.Get("/deis/registry/port", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	registry := host.Node.Value + ":" + port.Node.Value
```

Comment:
We've seen funny issues before if you don't check the value of `.Node`, as seen [here](https://github.com/deis/deis/commit/e803109cf9bc1bd868aab5c7d3bdcae18da589fd). That was only in the case for etcd's watch where they'd occasionally return a nil Node, but maybe it could happen during a `.Get` as well?